### ISSUE TITLE
BUG: RR scheduler may be trapped in long loop

### ISSUE THREAD

**Issue by hoang-tranviet** [multipath-tcp/mptcp#101]
In mptcp_rr.c:225, to check if a subflow is fully used:

```
if (rsp->quota == num_segments)
            full_subs++;
```

However, quota could surpass the num_segments:

```
if (skb->len > mss_now)
         rsp->quota += DIV_ROUND_UP(skb->len, mss_now);
```

This is pretty common at the beginning of the transfer. When this happens, the scheduler goes into a long loop until a timeout occurs for **every packet**. A simple fix could be  `if (rsp->quota >= num_segments)`
If you want, I can make a pull request for this.

**Comment by cpaasch** [comment]
Good catch! The test should probably be `if (rsp->quota >= num_segments)`. Feel free to send a patch.

**Comment by cpaasch** [comment]
Well, actually - while taking a closer look at the code, how can the infinite loop happen?
In order to loop, we have to `goto retry;`. But we only do this, when `iter == full_subs`. As quota is bigger than num_segments in the scenario you describe, full_subs won't be equal to iter.

Where you able to reproduce this infinite loop?

Irregardless of this, I agree that testing for >= num_segments is safer.

**Comment by hoang-tranviet** [comment]
Sure we do not have `iter == full_subs`. 
However, since the `choose_sk` is unset, after `goto found` the function always return NULL, and then the stack recalls scheduler again. So simply put, this is a bigger loop :)

**Comment by cpaasch** [comment]
Hmmm... I'm still not convinced that we can loop here. When the `next_segment` callback returns NULL, we stop looping in `mptcp_write_xmit()`.

I rather think that quota never gets down to zero again, and we thus will never select this subflow again. Please feel free to change the if and send out a patch. Thanks!

**Comment by hoang-tranviet** [comment]
The issue comes when the scheduler has fully filled all subflows. At this point, `full_subs` is smaller than `iter` for all subflows. As a result, even the valid subflows could not reset their quotas, not only the subflow with excessive quota.

**Comment by cpaasch** [comment]
Closing, as this is fixed by 8b98611bf0d15bc88b11632d34d294b0ccf739c4:

```
commit 8b98611bf0d15bc88b11632d34d294b0ccf739c4
Author: Viet-Hoang Tran <tranviet@uclouvain.be>
Date:   Wed Dec 2 05:25:59 2015 -0800

    mptcp: Fix quota checking in roundrobin scheduler

    Originally, the scheduler considers a subflow is fully used if
    rsp->quota == num_segments. However, this quota could surpass the
    num_segments if skb->len is larger than mss_now. When this happens,
    the quota will never be reset to zero.

    Signed-off-by: Viet-Hoang Tran <tranviet@uclouvain.be>
    Signed-off-by: Christoph Paasch <christoph.paasch@gmail.com>
```
